### PR TITLE
Refactor TLS1-PRF to create the MAC contexts early

### PR DESCRIPTION
**Pull request by levitte**
The TLS1-PRF implementation fetched the digest(s) for the underlying
MAC, just to get their names and pass those down to the MAC, which in
turn would fetch those same digests again.

This change circumvents this by fetching the MAC (or MACs in the
MD5-SHA1 special case) and create the MAC contexts for them directly
when this PRF receives the relevant parameters, thus only having to
pass EVP_MAC_CTX pointers around.

### CODE REVIEW

**Comment by levitte** [comment]
This is really a remnant from pre-3.0, where EVP_MDs and EVP_CIPHERs are passed around rather than their names.  We obviously sometimes still want to pass around EVP_MDs and EVP_CIPHERs, sometimes unnecessarily.  Our provider code could do well with a bit of extra scrutiny to see if we can catch this old habit.

File: providers/common/kdfs/tls1_prf.c [hunk 1]
```diff
@@ -276,23 +310,31 @@ static int tls1_prf_P_hash(const EVP_MD *md,
     unsigned char Ai[EVP_MAX_MD_SIZE];
     size_t Ai_len;
     int ret = 0;
-    OSSL_PARAM params[4];
+    OSSL_PARAM params[10], *p = params;
```
Comment (paulidale) [inline providers/common/kdfs/tls1_prf.c]: [6] is required here, not [10]

File: providers/common/kdfs/tls1_prf.c [hunk 2]
```diff
@@ -78,10 +79,18 @@ static int tls1_prf_alg(const EVP_MD *md, const EVP_MD *sha1,
 /* TLS KDF kdf context structure */
 typedef struct {
     void *provctx;
+
     /* Digest to use for PRF */
-    PROV_DIGEST digest;
+    char *digest;
     /* Second digest for the MD5/SHA-1 combined PRF */
-    PROV_DIGEST sha1;
+    const char *sha1;
```
Comment (paulidale) [inline providers/common/kdfs/tls1_prf.c]: Why?  This one is constant?

Resolving early is a performance gain.

**Review by paulidale — COMMENTED** [review]
I'm more leaning towards a `PROV_MAC` structure and associated helpers.  Early resolution of algorithms is good.

Other KDFs/PRFs would benefit from `PROC_MAC` (SSKDF uses a mac and KBKDF in #9924 does too).

I've likely missed some subtleties like I missed the `prov_ctx`.

**Comment by levitte** [comment]
> Early resolution of algorithms is good.

Well, in this case, they are getting "resolved", just to get "resolved" again by the MAC!  That's utterly ridiculous, and only a waste of cycles.

File: providers/common/kdfs/tls1_prf.c (hunk 2 above)
Comment (levitte) [reply providers/common/kdfs/tls1_prf.c]: Good question, it can be hard coded in `tls1_prf_alg`...

File: providers/common/kdfs/tls1_prf.c (hunk 1 above)
Comment (levitte) [reply providers/common/kdfs/tls1_prf.c]: Yeah, I set 10 to be safe before i started modifying and forgot to adjust.  Will fix.

**Comment by paulidale** [comment]
> Well, in this case, they are getting "resolved", just to get "resolved" again by the MAC! That's utterly ridiculous, and only a waste of cycles.

My poor choice of wording.  The MAC and any subservient algorithms should be resolved at _set_params time.

**Comment by levitte** [comment]
> My poor choice of wording. The MAC and any subservient algorithms should be resolved at `_set_params` time.

Aaahhhhhh!  Now that's a good idea!

I need a bit of time to get that right...

**Comment by levitte** [comment]
However, there's no need for a `PROV_MAC` structure since we don't need to deal with anything legacy there...

**Comment by paulidale** [comment]
> However, there's no need for a PROV_MAC structure since we don't need to deal with anything legacy there...

Do the underlying algorithms used alter this?
I.e. should the PROV_MAC structure include PROC_DIGEST and PROV_CIPHER fields for the MACS?
Perhaps a union of the two.

**Comment by levitte** [comment]
Hold ye' 'orses, @paulidale :wink:

**Comment by levitte** [comment]
Okie, I threw away the previous change and started over.  Please have a new look.

File: providers/common/kdfs/tls1_prf.c [hunk 3]
```diff
@@ -137,38 +138,73 @@ static int kdf_tls1_prf_derive(void *vctx, unsigned char *key,
...
     }
-    return tls1_prf_alg(md, ossl_prov_digest_md(&ctx->sha1),
+
+    return tls1_prf_alg(ctx->P_hash, ctx->P_sha1,
                         ctx->sec, ctx->seclen,
                         ctx->seed, ctx->seedlen,
                         key, keylen);
 }
 
+static EVP_MAC_CTX *kdf_tls1_prf_mkmac(OPENSSL_CTX *libctx,
```
Comment (t8m) [inline providers/common/kdfs/tls1_prf.c]: Nit: The mkmac name seems slightly confusing to me as it creates MAC_CTX and not the actual mac. So I'd prefer something like kdf_tls1_prf_macctx or kdf_tls1_prf_mkmacctx. 
But this is just for your consideration the approval holds even without the change.

**Review by t8m — APPROVED** [review]

File: providers/common/kdfs/tls1_prf.c (hunk 3 above)
Comment (levitte) [reply providers/common/kdfs/tls1_prf.c]: I agree with you, so I'll rename

**Comment by levitte** [comment]
Merged.

9a92bf1bffad15ede5ac97d1f1705c3e2c249a98 Refactor TLS1-PRF to create the MAC contexts early

**Comment by levitte** [comment]
Thanks all!

**Pull request closed without merging**